caproto/_commands.py

```diff
@@ -99,60 +99,28 @@ def padded_string_payload(payload):
-
-    if ntype != data_type:
-        size += (data_count - 1) * ctypes.sizeof(DBR_TYPES[ntype])
-
-    if data_count != 1:
-        # TODO this needs some work
-        assert data_count == len(values)
-        value_payload = b''.join(map(bytes, values))
+def data_payload(data, metadata, data_type, data_count):
+    if isinstance(data, ctypes.BigEndianStructure):
+        # This is already a DBR.
+        encoded_data = data
+    else:
+        # This is a tuple of values to be encoded into a DBR.
+        ntype = native_type(data_type)
+        cls_dict = {'_fields_': [('value', ntype * data_count)]}
+        cls = type(name, (ctypes.BigEndianStructure,), cls_dict)
+        cls(data)  # This is probably wrong in a super dumb way.
+    if isinstance(metamdata, ctypes.BigEndianStructure):
+        # This is already a DBR.
     else:
-        value_payload = DBR_TYPES[ntype](*values)
-
-    payload = DBR_TYPES[data_type]()
-
-    if metadata:
-        for attr in metadata_keywords:
-            # TODO note that some facilities use the nanosecond integer as
-            # a lossless event id and conversion to float is not a good thing
-            # ... should incorporate that into the logic here
-            if attr == 'timestamp':
-                if (hasattr(metadata, 'timestamp') and
-                        hasattr(payload, 'secondsSinceEpoch') and
-                        hasattr(payload, 'nanoSeconds')):
-                    sec, ns = timestamp_to_epics(metadata.timestamp)
-                    payload.secondsSinceEpoch = sec
-                    payload.nanoSeconds = ns
-                    continue
-
-            elif (attr == 'strs' and hasattr(metadata, attr) and
-                    hasattr(payload, attr)):
-                for i, string in enumerate(metadata.strs):
-                    bytes_ = string.encode('latin-1')
-                    justified = bytes_.ljust(MAX_ENUM_STRING_SIZE, b'\x00')
-                    payload.strs[i][:] = justified
-                payload.no_str = len(metadata.strs)
-                continue
-
-            if hasattr(metadata, attr) and hasattr(payload, attr):
-                value = getattr(metadata, attr)
-                if isinstance(value, str):
-                    value = value.encode('latin-1')
-
-                try:
-                    setattr(payload, attr, value)
-                except Exception as ex:
-                    # TODO server probably should not fail here
-                    # raise ValueError('Invalid metadata for {}={} ({})'
-                    #                  ''.format(attr, value, ex))
-                    print('set metadata fail', attr, value)
-    payload.value = value_payload.value
-
-    payload = bytes(payload)
+        # This is a tuple of values to be encoded into a DBR.
+        justified_md = []
+        for val in metadata:
+            if isinstance(val, str):
+                val = bytes_(val).ljust(MAX_ENUM_STRING_SIZE, b'\x00')
```

Comment:
`len(val)` could be > `MAX_ENUM_STRING_SIZE`, then if so I guess the DBR_TYPES[]() initializer bails or chops off the rest?